velox/exec/Task.cpp

```diff
@@ -418,6 +420,50 @@ void Task::removeSpillDirectoryIfExists() {
   }
 }
 
+/// Builds the query trace config.
+std::optional<QueryTraceConfig> Task::maybeMakeTraceConfig() const {
+  const auto& queryConfig = queryCtx_->queryConfig();
+  if (!queryConfig.queryTraceEnabled()) {
+    return std::nullopt;
+  }
+
+  const auto traceDir = queryConfig.queryTraceDir();
+  VELOX_CHECK(
+      !traceDir.empty(), "Query trace enabled but the trace dir is not set.");
+
+  const auto queryTraceNodes = queryConfig.queryTraceNodes();
+  if (queryTraceNodes.empty()) {
+    LOG(WARNING) << "Query trace enabled but no tracing plan node specified.";
+    return std::nullopt;
+  }
+
+  std::vector<std::string> nodes;
+  folly::split(',', queryTraceNodes, nodes);
+  const std::unordered_set<std::string> nodeSet(nodes.begin(), nodes.end());
+  VELOX_CHECK_EQ(nodeSet.size(), nodes.size());
+
+  const auto writeBufferSize = queryConfig.queryWriteBufferSize();
+
+  return QueryTraceConfig(nodeSet, traceDir, writeBufferSize);
+}
+
+std::string Task::createTraceDirectory(const std::string& traceDir) const {
+  std::lock_guard<std::mutex> l(fileOpMutex_);
```

Comment:
We probably doesn't need the lock as there is not conflict. Also first remove the directory and then create the new one.